### ISSUE TITLE
Improve "brimcap analyze" error messaging & diagnostics

### ISSUE THREAD

**Issue by philrz** [brimdata/brimcap#16]
Since part of what we're trying to achieve with Brimcap is to make it easier for users to bring their own custom Zeek/Suricata, it'll be important that we surface details with problems in such tools when they're invoked via `brimcap analyze` during pcap processing.

To cite a specific example, here's what a macOS user sees at the moment with `brimcap` commit `688a7f5` and Homebrew-installed Zeek v4.0.0 and Suricata v6.0.2 when analyzing [https://archive.wrccdc.org/pcaps/2018/wrccdc.2018-03-23.010014000000000.pcap.gz:](https://archive.wrccdc.org/pcaps/2018/wrccdc.2018-03-23.010014000000000.pcap.gz:)

```
$ brimcap analyze -z ../wrccdc.pcap > /dev/null
1 warnings occurred while parsing log data:
    /var/folders/yn/jbkxxkpd4vg142pc3_bd_krc0000gn/T/brimcap-2408807547/eve.json: duplicate field subject: x1
```

There are known issues at play here, specifically that Suricata v6 is generating bad output (#13) and Zed could stand to provide more detail about where the "duplicate field" problem arose in the data (brimdata/zed#2452). But putting those aside, some things that would have improved the user experience if it had arisen in the wild:

1.  Provide a way for the user to see the stdout and stderr of the Zeek/Suricata processes invoked via `brimcap analyze` so the user knows whether those pieces ran healthy. In this case they did, but users might encounter any number of speedbumps if they're iteratively tweaking configs or adding customizations to these systems.
2.  Make it clear if the choke happened downstream of the invoked processes, such as in this case where it happened in the Zed shaping step.
3.  Have the ability to leave behind the logs generated by Zeek/Suricata so the user can examine them. In this case, the user could have used `zq` or other JSON processing tools on the raw EVE JSON output to isolate which lines had the duplicate field.
4.  Show the command lines that `suricata analyze` used when invoking these processes, so the user can easily repeat this in their debug steps.

In summary, I expect our `brimcap analyze` support statement will be something like "if it doesn't 'just work', repeat the individual steps performed by `brimcap analyze` and see where it goes wrong". In this regard, having a debug mode that's as transparent as possible will be very helpful.

**Comment by mattnibs** [comment]
I support the overall gist of this issue, but I don't think any of this would help this issue:

```
brimcap analyze -z ../wrccdc.pcap > /dev/null
1 warnings occurred while parsing log data:
    /var/folders/yn/jbkxxkpd4vg142pc3_bd_krc0000gn/T/brimcap-2408807547/eve.json: duplicate field subject: x1
```

Suricata doesn't provide any notable debug output from this. I suppose having the ability to leave behind the dir that brimcap creates would at least allow someone to inspect the data left behind.

**Comment by philrz** [comment]
@mattnibs: A bit more debug output might not get them _directly_ to the [root cause Suricata issue](https://redmine.openinfosecfoundation.org/issues/4106), but I think it still would have helped get them further along. Even if the debug output doesn't show a smoking gun, being able to see it to rule out common problems is helpful, such as by comparing it side-by-side next to outputs when processing pcaps that may not have exhibited the same problem. Revisiting my points above:

1.  If the user was familiar with what a healthy Suricata run looks like, allowing them to see Suricata's stderr/stdout might have helped them determine if a problem might have been Suricata's fault. For instance, one of the things I've noticed Suricata can do a lot when its memory limits are set too low is it'll enter a "Flow emergency mode" where it starts flushing its connection state tables during packet processing and producing bad output in the EVE JSON. When I first saw this symptom the first thing I wanted to know was whether Suricata had entered this mode while processing the pcap, since I've been burned by it before.
2.  It would have helped for the user to know the "duplicate field" warning came from Zed processing of the EVE JSON. It's kind of implied in the output above and of course obvious to those of us who know the pieces, but maybe not so much for a newbie (e.g. they might not even know EVE is the name of Suricata's output format). I suppose we could debate whether the default output should be terse or verbose, but given that there's multiple steps, when a Brimcap user needs to debug a problem having it broken out with explicitly labeled parts for the Zeek/Suricata parts vs. the parsing into ZNG part would help to divide & conquer.

...and I think you've stated agreement with my 3rd item, & my 4th point stands as originally written.

**Comment by mattnibs** [comment]
For 4, I will say that for now, suricatarunner/zeekrunner obscures how zeek/suricata are actually run so we can hold off on this to now.

**Comment by philrz** [comment]
That's fair point, though in that case showing the command that was used to invoke the runner would be fair game so that if they're unfamiliar with how the runner script invokes the tools, they can start to look into those.

Part of me is still planning for the likelihood that we'll eventually offer the option for users to point at arbitrary custom Zeek/Suricata installs in their path. Like you say, we can revisit that subject at the appropriate time. I know we need to get the basic end-to-end working before anything else.

**Comment by philrz** [comment]
Here's a verification of the change in #26 using `brimcap` commit `1fa5fc4`. For test data I'm using the attached [ssl.pcapng.gz](https://github.com/brimdata/brimcap/files/6339514/ssl.pcapng.gz) (uncompressed), which is a flow extracted from https://archive.wrccdc.org/pcaps/2018/wrccdc.2018-03-23.010014000000000.pcap.gz.

Let's say I'm a user that ultimately wants to try using my own custom Zeek and Suricata that are in my `$PATH`.

```
$ zeek --version
zeek version 4.0.0

$ suricata -V
This is Suricata version 6.0.2 RELEASE
```

I start by getting a baseline using the bundled Zeek & Suricata artifacts. This shows me some Zeek events and a Suricata alert were generated.

```
$ brimcap analyze -Z ssl.pcapng > ssl-builtin.zson
$ zq -z 'cut _path,event_type' ssl-builtin.zson
{_path:"stats"}
{_path:"stats"}
{_path:"capture_loss"}
{_path:"files"}
{_path:"files"}
{_path:"files"}
{_path:"x509"}
{_path:"ssl"}
{_path:"conn"}
{event_type:"alert" (bstring)} (=0)
```

Since Brimcap expects its analyzers to take input on stdin, I create simple wrapper scripts to expect this while invoking the ones in my path.

```
$ cat /tmp/myzeek 
#!/bin/bash
zeek -C -r - --exec "event zeek_init() { Log::disable_stream(PacketFilter::LOG); Log::disable_stream(LoadedScripts::LOG); }" local

$ cat /tmp/mysuricata 
#!/bin/bash
suricata -r /dev/stdin
```

Then I create a simple config YAML and try to use it to invoke my wrapper scripts.

```
$ cat simple.yml 
analyzers:
  - cmd: /tmp/myzeek
  - cmd: /tmp/mysuricata

$ brimcap analyze -Z -config simple.yml ssl.pcapng > ssl-simple.zson
{"type":"error","error":"duplicate field subject"}
```

This error message is not ideal because it doesn't explain at which phase the error happened (Zeek? Suricata? Zed processing of the logs they generated?) but it does tell me there's a problem. Here I add the `workdir` parameter to my YAML so I can see what logs were generated by each process. Note that these directories must exist, and any existing contents in them (i.e. from a prior run) are _not_ deleted by Brimcap, so you may want to clear them out between runs.

```
$ cat simple-workdir.yml 
analyzers:
  - cmd: /tmp/myzeek
    workdir: myzeek-wd
  - cmd: /tmp/mysuricata
    workdir: mysuricata-wd

$ mkdir myzeek-wd
$ mysuricata-wd

$ brimcap analyze -Z -config simple-workdir.yml ssl.pcapng > ssl-simple-workdir.zson
{"type":"error","error":"duplicate field subject"}
```

Now I can try running `zq` directly on the logs and narrow down the problem.

```
$ cd myzeek-wd/
$ ls -l
total 48
-rw-r--r--  1 phil  staff   278 Apr 19 16:07 capture_loss.log
-rw-r--r--  1 phil  staff   626 Apr 19 16:07 conn.log
-rw-r--r--  1 phil  staff  1417 Apr 19 16:07 files.log
-rw-r--r--  1 phil  staff   910 Apr 19 16:07 ssl.log
-rw-r--r--  1 phil  staff   787 Apr 19 16:07 stats.log
-rw-r--r--  1 phil  staff  1023 Apr 19 16:07 x509.log

$ zq -z 'count() by _path' *
{_path:"files",count:3 (uint64)} (=0)
{_path:"ssl",count:1} (0)
{_path:"stats",count:2} (0)
{_path:"x509",count:1} (0)
{_path:"capture_loss",count:1} (0)
{_path:"conn",count:1} (0)

$ cd ../mysuricata-wd/
$ ls -l
total 48
-rw-r--r--  1 phil  staff  8401 Apr 19 16:07 eve.json
-rw-r--r--  1 phil  staff   202 Apr 19 16:07 fast.log
-rw-r--r--  1 phil  staff  2539 Apr 19 16:07 stats.log
-rw-r--r--  1 phil  staff  1387 Apr 19 16:07 suricata.log

$ zq -z 'count() by event_type' eve.json 
eve.json: duplicate field subject
```

We then recognize this as an incident of https://redmine.openinfosecfoundation.org/issues/4106. While we're waiting on a Zed fix in https://github.com/brimdata/zed/issues/2523, we can work around the problem by post-processing the data through `jq`, since it happens to deduplicate JSON. However, looking at the contents of the directory also makes us realize that we only want Brimcap to work on the deduplicated output and not the original `eve.json`, so we also use the `globs` parameter in the config YAML to target that one file.

```
$ cat /tmp/mysuricata 
#!/bin/bash
suricata -r /dev/stdin
cat eve.json | jq -c . > deduped-eve.json

$ cat simple-workdir-deduped.yml
analyzers:
  - cmd: /tmp/myzeek
    workdir: myzeek-wd
  - cmd: /tmp/mysuricata
    workdir: mysuricata-wd
    globs: ["deduped-eve.json"]

$ brimcap analyze -Z -config simple-workdir-deduped.yml ssl.pcapng > ssl-simple-workdir-deduped.zson

$ zq -z 'cut _path,event_type' ssl-simple-workdir-deduped.zson
{_path:"stats"}
{_path:"stats"}
{_path:"capture_loss"}
{_path:"files"}
{_path:"files"}
{_path:"files"}
{_path:"x509"}
{_path:"ssl"}
{_path:"conn"}
{event_type:"tls"}
{event_type:"alert"}
{event_type:"flow"}
{event_type:"stats"}
```

That worked! Of course, the `zq` output also reminds us that my custom Suricata has excess events beyond just the alerts that come out of the one bundled with Brimcap, so we may want to think about trimming those by applying a shaper. But we're on our way with a debug workflow.

Thanks @mattnibs!
